File: src/main/scala/com/xmlcalabash/functions/UrifyFunction.scala
package com.xmlcalabash.functions

import com.xmlcalabash.config.XMLCalabashConfig
import com.xmlcalabash.exceptions.XProcException
import com.xmlcalabash.messages.XProcItemMessage
import com.xmlcalabash.model.util.XProcConstants
import com.xmlcalabash.util.{URIUtils, Urify}
import net.sf.saxon.expr.{Expression, StaticContext, XPathContext}
import net.sf.saxon.functions.AccessorFn.Component
import net.sf.saxon.lib.{ExtensionFunctionCall, ExtensionFunctionDefinition}
import net.sf.saxon.om.{Item, Sequence, StructuredQName}
import net.sf.saxon.s9api.{QName, XdmAtomicValue, XdmEmptySequence, XdmItem, XdmNode, XdmValue}
import net.sf.saxon.tree.iter.ArrayIterator
import net.sf.saxon.tree.tiny.{TinyDocumentImpl, TinyElementImpl, TinyTextImpl}
import net.sf.saxon.value.{QNameValue, SequenceType, StringValue}

class UrifyFunction(runtime: XMLCalabashConfig) extends FunctionImpl() {
  val funcname = new StructuredQName("p", XProcConstants.ns_p, "urify")

  override def getFunctionQName: StructuredQName = funcname

  override def getArgumentTypes: Array[SequenceType] = Array(SequenceType.SINGLE_ITEM, SequenceType.OPTIONAL_ITEM)

  override def getResultType(suppliedArgumentTypes: Array[SequenceType]): SequenceType = SequenceType.SINGLE_ITEM

  override def makeCallExpression(): ExtensionFunctionCall = {
    new UrifyCall(this)
  }

  private class UrifyCall(funcdef: ExtensionFunctionDefinition) extends ExtensionFunctionCall {
    private var staticContext: StaticContext = _

    override def supplyStaticContext(context: StaticContext, locationId: Int, arguments: Array[Expression]): Unit = {
      super.supplyStaticContext(context, locationId, arguments)
      staticContext = context
    }

    override def call(context: XPathContext, arguments: Array[Sequence]): Sequence = {
      val exprEval = runtime.expressionEvaluator
      if (exprEval.dynContext.isEmpty) {
        throw XProcException.xiExtFunctionNotAllowed()
      }

      val relativeuri = arguments(0).head.getStringValue
      val proposedbase = arguments(1).head.getStringValue

      if (proposedbase == "") {
        throw XProcException.xdUrifyFailed(relativeuri, proposedbase, None)
      }

      val result = new Urify(proposedbase).resolve(relativeuri)
      new StringValue(result.toString)
    }
  }
}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/functions/UrifyFunction.kt
package com.xmlcalabash.functions

import com.xmlcalabash.config.SaxonConfiguration
import com.xmlcalabash.namespace.NsP
import com.xmlcalabash.util.Urify
import net.sf.saxon.expr.XPathContext
import net.sf.saxon.lib.ExtensionFunctionCall
import net.sf.saxon.lib.ExtensionFunctionDefinition
import net.sf.saxon.om.Sequence
import net.sf.saxon.om.StructuredQName
import net.sf.saxon.value.SequenceType
import net.sf.saxon.value.StringValue

class UrifyFunction(private val config: SaxonConfiguration): ExtensionFunctionDefinition() {
    override fun getFunctionQName(): StructuredQName {
        return StructuredQName("p", NsP.namespace, "urify")
    }

    override fun getArgumentTypes(): Array<SequenceType> {
        return arrayOf(SequenceType.SINGLE_ITEM, SequenceType.OPTIONAL_ITEM)
    }

    override fun getResultType(suppliedArgumentTypes: Array<out SequenceType>?): SequenceType {
        return SequenceType.SINGLE_ITEM
    }

    override fun makeCallExpression(): ExtensionFunctionCall {
        return UrifyImpl()
    }

    inner class UrifyImpl: ExtensionFunctionCall() {
        override fun call(context: XPathContext?, arguments: Array<out Sequence>?): Sequence {
            val relativeUri = arguments!![0].head().stringValue
            if (arguments.size > 1 && arguments[1].head() != null) {
                return StringValue(Urify.urify(relativeUri, arguments[1].head().stringValue))
            }
            return StringValue(Urify.urify(relativeUri))
        }
    }
}